python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -0,0 +1,226 @@
+    register_model,
+    type_callable,
+    typeof_impl,
+)
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AbstractTemplate, AttributeTemplate
+from numba.cuda.cudadecl import registry as cuda_registry
+from numba.np import numpy_support
+
+# Disable occupancy warnings to avoid polluting output when there are few
+# groups.
+numba.config.CUDA_LOW_OCCUPANCY_WARNINGS = 0
+
+index_default_type = types.int64
+SUPPORTED_GROUPBY_NUMBA_TYPES = [types.int64, types.float64]
+SUPPORTED_GROUPBY_NUMPY_TYPES = [
+    numpy_support.as_dtype(dt) for dt in SUPPORTED_GROUPBY_NUMBA_TYPES
+]
+
+
+class Group(object):
+    def __init__(self, group_data, size, index, dtype, index_dtype):
+        self.group_data = group_data
+        self.size = size
+        self.index = index
+        self.dtype = dtype
+        self.index_dtype = index_dtype
+
+
+class GroupType(numba.types.Type):
+    def __init__(self, group_scalar_type, index_type=index_default_type):
+        self.group_scalar_type = group_scalar_type
+        self.index_type = index_type
+        self.group_data_type = types.CPointer(group_scalar_type)
+        self.size_type = types.int64
+        self.group_index_type = types.CPointer(index_type)
+        super().__init__(
+            name=f"Group({self.group_scalar_type}, {self.index_type})"
+        )
+
+
+@typeof_impl.register(Group)
+def typeof_group(val, c):
+    return GroupType(
+        numba.np.numpy_support.from_dtype(val.dtype),
+        numba.np.numpy_support.from_dtype(val.index_dtype),
+    )  # Identifies instances of the Group class as GroupType
+
+
+# The typing of the python "function" Group.__init__
+# as it appears in python code
+@type_callable(Group)
+def type_group(context):
+    def typer(group_data, size, index):
+        if (
+            isinstance(group_data, types.Array)
+            and isinstance(size, types.Integer)
+            and isinstance(index, types.Array)
+        ):
+            return GroupType(group_data.dtype, index.dtype)
+
+    return typer
+
+
+@register_model(GroupType)
+class GroupModel(models.StructModel):
+    def __init__(
+        self, dmm, fe_type
+    ):  # fe_type is fully instantiated group type
+        members = [
+            ("group_data", types.CPointer(fe_type.group_scalar_type)),
+            ("size", types.int64),
+            ("index", types.CPointer(fe_type.index_type)),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+call_cuda_functions: Dict[Any, Any] = {}
+
+
+def _register_cuda_reduction_caller(func, inputty, retty):
+    cuda_func = cuda.declare_device(
+        f"Block{func}_{inputty}", retty(types.CPointer(inputty), types.int64)
+    )
+
+    def caller(data, size):
+        return cuda_func(data, size)
+
+    if call_cuda_functions.get(func.lower()) is None:
+        call_cuda_functions[func.lower()] = {}
+
+    type_key = (retty, inputty)
+    call_cuda_functions[func.lower()][type_key] = caller
```

Comment:
Why can you not just use `cuda_func` directly here?

Reply:
This has to do with hammering what comes out of `cuda.declare_device` into what `compile_internal` expects. The former returns a 

`cuda.declare_device` returns a `numba.cuda.compiler.ExternFunction` but later on `compile_internal` expects a standard python callable with `__code__` attributes and so forth.